### PR TITLE
Make Object::is_a faster

### LINKED ISSUE #1874: Object#is_a? is not very efficient
The current code 
```c++
bool Object::is_a(Env *env, Value val) const {
   if (/* direct match (instance_of?), not relevant here */) {
    } else {
        ClassObject *klass = singleton_class();
        if (!klass)
            klass = m_klass;
        ArrayObject *ancestors = klass->ancestors(env);
        for (Value m : *ancestors) {
            if (module == m->as_module()) {
                return true;
            }
        }
        return false;
    }
}
```
In other words: unless we have a direct match, we first build a list of all the ancestors, then loop through the results to see if the wanted value is in it.

This could be done more efficient by removing the temporary array and directly looping with a method. The current code is pretty much like `enumerable.to_a.include?(val)`, which could be written as `enumerable.include?(val)`.

### PR DESCRIPTION
**Pull request by tekknolagi**
Don't allocate an intermediate ArrayObject.

Fix #1874

### CODE REVIEW

**Comment by tekknolagi** [comment]
I'm not sure if I need the lock anymore since I am not allocating.

**Comment by herwinw** [comment]
> I'm not sure if I need the lock anymore since I am not allocating.

That one should no longer be needed. The code is very similar to `bool ModuleObject::is_subclass_of(ModuleObject *other)` just below your new code, and that one doesn't require it either.

And to be forward about this: my (undocumented) end goal here was to create an iterable interface for the ancestors. We now have `ModuleObject::ancestors` and `ModuleObject::ancestors_include`, so I kind of want to use things like `std::any_of` for that second one.
This does not mean that I don't like your change. Quite the opposite: if I were to implement the iterator interface from the current master, I would probably start with a commit like this one just to illustrate the similarities between these two methods, to make it easier to extract the common code.

File: src/module_object.cpp
Comment (herwinw) [inline src/module_object.cpp:666]: As described in a bit more detail in the comments: this line can be removed. Once it is gone, I would say it's good to merge.

**Review by herwinw — APPROVED** [review]

File: src/module_object.cpp
Comment (seven1m) [reply src/module_object.cpp]: I'm going to investigate this more in #1956. My gut tells me we *do* want the lock here (and in the method below). But I won't hold up this work!

**Review by seven1m — APPROVED** [review]
Thanks!

File: src/module_object.cpp
Comment (tekknolagi) [reply src/module_object.cpp]: Oh, why's that?

**Pull request merged**
